### PR TITLE
Move to JSON Logging complete!

### LINKED ISSUE #221: Move logging system to JSON
Convert the logging system we have now to the proposed schema from #194

### PR DESCRIPTION
**Pull request by VivaanBahl**
The final version is still fully untested (GPS recently got a quikfix and unit tests passed) but all the other components are fully tested.
The logs are now in JSON format, which follows the example in log_schema.txt and log_schema_example.txt, found on the vrbahl_logging branch

Closes #221

### CODE REVIEW

**Comment by TrevorDecker** [comment]
Awesome! Test the gps before we merge this into master.  Also it appears that you have conflicts with master, you should fix that.

**Comment by VivaanBahl** [comment]
@TrevorDecker OK, I managed to move stuff around to make it work with Vasu's new sensor setup, but I feel like we should probably schedule another full test at some point.

**Comment by TrevorDecker** [comment]
OK lets test it on Thursday, also if you get free time  before hand feel free to take the buggy out for testing.

File: real_time/surface_src/java_src/Alice/src/main/java/com/roboclub/robobuggy/logging/SensorLogger.java [hunk 1]
```diff
@@ -9,15 +9,24 @@
 import java.util.Queue;
 import java.util.concurrent.LinkedBlockingQueue;
 
+import com.orsoncharts.util.json.JSONObject;
+import com.roboclub.robobuggy.main.config;
```
Comment (TrevorDecker) [inline real_time/surface_src/java_src/Alice/src/main/java/com/roboclub/robobuggy/logging/SensorLogger.java]: Codacy Issue found: [Avoid unused imports such as 'com.roboclub.robobuggy.main.config'](https://www.codacy.com/app/Mr-Trevor-Decker/RoboBuggy/file/1263455875/issues/source?bid=2938665&fileBranchId=2938743#l13)

File: real_time/surface_src/java_src/Alice/src/main/java/com/roboclub/robobuggy/main/config.java [hunk 2]
```diff
@@ -26,6 +26,7 @@
 	// "C:\\Users\\abc\\buggy-log\\VisionSystem.exe";
 
 	public static String LOG_FILE_LOCATION = "logs";
+	public static String LOG_STOP_MESSAGE = "STOP_LOGGING";
```
Comment (TrevorDecker) [inline real_time/surface_src/java_src/Alice/src/main/java/com/roboclub/robobuggy/main/config.java]: Codacy Issue found: [The field name indicates a constant but its modifiers do not](https://www.codacy.com/app/Mr-Trevor-Decker/RoboBuggy/file/1263455880/issues/source?bid=2938665&fileBranchId=2938743#l29)

File: real_time/surface_src/java_src/Alice/src/main/java/com/roboclub/robobuggy/main/mainFile.java [hunk 3]
```diff
@@ -60,5 +60,6 @@ public static void bringup_sim() throws Exception {
         String ImuKey = sm.newRealSensor(RealNodeEnum.IMU, config.COM_PORT_IMU, SensorChannel.IMU);
         String GpsKey = sm.newRealSensor(RealNodeEnum.GPS, config.COM_PORT_GPS_INTEGRATED, SensorChannel.GPS);
         String RBSMKey = sm.newRealSensor(RealNodeEnum.RBSM, config.COM_PORT_ENCODER, SensorChannel.ENCODER, SensorChannel.STEERING);
+        String LoggingKey = sm.newRealSensor(RealNodeEnum.LOGGING_BUTTON, "", SensorChannel.GUI_LOGGING_BUTTON);
```
Comment (TrevorDecker) [inline real_time/surface_src/java_src/Alice/src/main/java/com/roboclub/robobuggy/main/mainFile.java]: Codacy Issue found: [Avoid unused local variables such as 'LoggingKey'.](https://www.codacy.com/app/Mr-Trevor-Decker/RoboBuggy/file/1263455881/issues/source?bid=2938665&fileBranchId=2938743#l63)

File: real_time/surface_src/java_src/Alice/src/main/java/com/roboclub/robobuggy/nodes/GpsNode.java [hunk 4]
```diff
@@ -3,6 +3,9 @@
 import java.text.SimpleDateFormat;
 import java.util.Date;
 
+import jdk.nashorn.internal.scripts.JS;
```
Comment (TrevorDecker) [inline real_time/surface_src/java_src/Alice/src/main/java/com/roboclub/robobuggy/nodes/GpsNode.java]: Codacy Issue found: [Avoid unused imports such as 'jdk.nashorn.internal.scripts.JS'](https://www.codacy.com/app/Mr-Trevor-Decker/RoboBuggy/file/1263455904/issues/source?bid=2938665&fileBranchId=2938743#l6)

File: real_time/surface_src/java_src/Alice/src/main/java/com/roboclub/robobuggy/nodes/RBSMNode.java [hunk 5]
```diff
@@ -144,4 +145,35 @@ public int peel(byte[] buffer, int start, int bytes_available) {
...
+		JSONObject data = new JSONObject();
+		JSONObject params = new JSONObject();
+		
+		data.put("timestamp", messageData[1]);
+		
+		if(sensorName.equals("steering")) {
+			data.put("name", "Steering");
+			params.put("angle", Float.valueOf(messageData[2]));
+		}
+		else if (sensorName.equals("encoder")) {
```
Comment (TrevorDecker) [inline real_time/surface_src/java_src/Alice/src/main/java/com/roboclub/robobuggy/nodes/RBSMNode.java]: Codacy Issue found: [Position literals first in String comparisons](https://www.codacy.com/app/Mr-Trevor-Decker/RoboBuggy/file/1263455908/issues/source?bid=2938665&fileBranchId=2938743#l163)

File: real_time/surface_src/java_src/Alice/src/main/java/com/roboclub/robobuggy/logging/SensorLogger.java (hunk 1 above)
Comment (TrevorDecker) [inline real_time/surface_src/java_src/Alice/src/main/java/com/roboclub/robobuggy/logging/SensorLogger.java]: Codacy Issue found: [Avoid unused imports such as 'com.roboclub.robobuggy.main.config'](https://www.codacy.com/app/Mr-Trevor-Decker/RoboBuggy/file/1272370353/issues/source?bid=2938665&fileBranchId=2938743#l13)

File: real_time/surface_src/java_src/Alice/src/main/java/com/roboclub/robobuggy/main/config.java (hunk 2 above)
Comment (TrevorDecker) [inline real_time/surface_src/java_src/Alice/src/main/java/com/roboclub/robobuggy/main/config.java]: Codacy Issue found: [The field name indicates a constant but its modifiers do not](https://www.codacy.com/app/Mr-Trevor-Decker/RoboBuggy/file/1272370358/issues/source?bid=2938665&fileBranchId=2938743#l29)

File: real_time/surface_src/java_src/Alice/src/main/java/com/roboclub/robobuggy/main/mainFile.java (hunk 3 above)
Comment (TrevorDecker) [inline real_time/surface_src/java_src/Alice/src/main/java/com/roboclub/robobuggy/main/mainFile.java]: Codacy Issue found: [Avoid unused local variables such as 'LoggingKey'.](https://www.codacy.com/app/Mr-Trevor-Decker/RoboBuggy/file/1272370359/issues/source?bid=2938665&fileBranchId=2938743#l63)

File: real_time/surface_src/java_src/Alice/src/main/java/com/roboclub/robobuggy/nodes/GpsNode.java (hunk 4 above)
Comment (TrevorDecker) [inline real_time/surface_src/java_src/Alice/src/main/java/com/roboclub/robobuggy/nodes/GpsNode.java]: Codacy Issue found: [Avoid unused imports such as 'jdk.nashorn.internal.scripts.JS'](https://www.codacy.com/app/Mr-Trevor-Decker/RoboBuggy/file/1272370381/issues/source?bid=2938665&fileBranchId=2938743#l6)

File: real_time/surface_src/java_src/Alice/src/main/java/com/roboclub/robobuggy/nodes/RBSMNode.java (hunk 5 above)
Comment (TrevorDecker) [inline real_time/surface_src/java_src/Alice/src/main/java/com/roboclub/robobuggy/nodes/RBSMNode.java]: Codacy Issue found: [Position literals first in String comparisons](https://www.codacy.com/app/Mr-Trevor-Decker/RoboBuggy/file/1272370385/issues/source?bid=2938665&fileBranchId=2938743#l163)

File: real_time/surface_src/java_src/Alice/src/main/java/com/roboclub/robobuggy/main/config.java [hunk 6]
```diff
@@ -26,6 +26,23 @@
...
+	public static String SW_VERSION_HIGHLEVEL = "1.0.0";
+	
+	
+	//names of the various sensors
+	public static String SENSOR_NAME_BRAKE = "brake";
+	public static String SENSOR_NAME_CAMERA = "camera";
+	public static String SENSOR_NAME_ENCODER = "encoder";
+	public static String SENSOR_NAME_GPS = "gps";
+	public static String SENSOR_NAME_GUI_LOGGING_BUTTON = "logging_button";
+	public static String SENSOR_NAME_IMU = "imu";
```
Comment (TrevorDecker) [inline real_time/surface_src/java_src/Alice/src/main/java/com/roboclub/robobuggy/main/config.java]: Codacy Issue found: [The field name indicates a constant but its modifiers do not](https://www.codacy.com/app/Mr-Trevor-Decker/RoboBuggy/file/1273020439/issues/source?bid=2938665&fileBranchId=2938743#l39)

File: real_time/surface_src/java_src/Alice/src/main/java/com/roboclub/robobuggy/main/config.java [hunk 7]
```diff
@@ -26,6 +26,23 @@
 	// "C:\\Users\\abc\\buggy-log\\VisionSystem.exe";
 
 	public static String LOG_FILE_LOCATION = "logs";
+	public static String LOG_STOP_MESSAGE = "STOP_LOGGING";
```
Comment (TrevorDecker) [inline real_time/surface_src/java_src/Alice/src/main/java/com/roboclub/robobuggy/main/config.java]: Codacy Issue found: [The field name indicates a constant but its modifiers do not](https://www.codacy.com/app/Mr-Trevor-Decker/RoboBuggy/file/1273020439/issues/source?bid=2938665&fileBranchId=2938743#l29)

File: real_time/surface_src/java_src/Alice/src/main/java/com/roboclub/robobuggy/main/config.java [hunk 8]
```diff
@@ -26,6 +26,23 @@
...
+	public static String SW_VERSION_HIGHLEVEL = "1.0.0";
+	
+	
+	//names of the various sensors
+	public static String SENSOR_NAME_BRAKE = "brake";
+	public static String SENSOR_NAME_CAMERA = "camera";
+	public static String SENSOR_NAME_ENCODER = "encoder";
+	public static String SENSOR_NAME_GPS = "gps";
```
Comment (TrevorDecker) [inline real_time/surface_src/java_src/Alice/src/main/java/com/roboclub/robobuggy/main/config.java]: Codacy Issue found: [The field name indicates a constant but its modifiers do not](https://www.codacy.com/app/Mr-Trevor-Decker/RoboBuggy/file/1273020439/issues/source?bid=2938665&fileBranchId=2938743#l37)

File: real_time/surface_src/java_src/Alice/src/main/java/com/roboclub/robobuggy/main/mainFile.java (hunk 3 above)
Comment (TrevorDecker) [inline real_time/surface_src/java_src/Alice/src/main/java/com/roboclub/robobuggy/main/mainFile.java]: Codacy Issue found: [Avoid unused local variables such as 'LoggingKey'.](https://www.codacy.com/app/Mr-Trevor-Decker/RoboBuggy/file/1273020440/issues/source?bid=2938665&fileBranchId=2938743#l63)

File: real_time/surface_src/java_src/Alice/src/main/java/com/roboclub/robobuggy/nodes/GpsNode.java (hunk 4 above)
Comment (TrevorDecker) [inline real_time/surface_src/java_src/Alice/src/main/java/com/roboclub/robobuggy/nodes/GpsNode.java]: Codacy Issue found: [Avoid unused imports such as 'jdk.nashorn.internal.scripts.JS'](https://www.codacy.com/app/Mr-Trevor-Decker/RoboBuggy/file/1273020462/issues/source?bid=2938665&fileBranchId=2938743#l6)

File: real_time/surface_src/java_src/Alice/src/main/java/com/roboclub/robobuggy/main/config.java [hunk 9]
```diff
@@ -26,6 +26,23 @@
...
+	public static String SENSOR_NAME_GUI_LOGGING_BUTTON = "logging_button";
+	public static String SENSOR_NAME_IMU = "imu";
+	public static String SENSOR_NAME_STEERING = "steering";
+	
+	//other message types
+	public static String MESSAGE_TYPE_RESET = "reset";
+	public static String MESSAGE_TYPE_STATUS = "status";
```
Comment (TrevorDecker) [inline real_time/surface_src/java_src/Alice/src/main/java/com/roboclub/robobuggy/main/config.java]: Codacy Issue found: [The field name indicates a constant but its modifiers do not](https://www.codacy.com/app/Mr-Trevor-Decker/RoboBuggy/file/1273020439/issues/source?bid=2938665&fileBranchId=2938743#l44)

File: real_time/surface_src/java_src/Alice/src/main/java/com/roboclub/robobuggy/nodes/ImuNode.java [hunk 10]
```diff
@@ -1,7 +1,10 @@
 package com.roboclub.robobuggy.nodes;
 
+import com.orsoncharts.util.json.JSONObject;
+import com.roboclub.robobuggy.logging.IMUMeasurement;
```
Comment (TrevorDecker) [inline real_time/surface_src/java_src/Alice/src/main/java/com/roboclub/robobuggy/nodes/ImuNode.java]: Codacy Issue found: [Avoid unused imports such as 'com.roboclub.robobuggy.logging.IMUMeasurement'](https://www.codacy.com/app/Mr-Trevor-Decker/RoboBuggy/file/1273313808/issues/source?bid=2938665&fileBranchId=2938743#l4)

File: real_time/surface_src/java_src/Alice/src/main/java/com/roboclub/robobuggy/main/config.java (hunk 7 above)
Comment (TrevorDecker) [inline real_time/surface_src/java_src/Alice/src/main/java/com/roboclub/robobuggy/main/config.java]: Codacy Issue found: [The field name indicates a constant but its modifiers do not](https://www.codacy.com/app/Mr-Trevor-Decker/RoboBuggy/file/1273313784/issues/source?bid=2938665&fileBranchId=2938743#l29)

File: real_time/surface_src/java_src/Alice/src/main/java/com/roboclub/robobuggy/main/mainFile.java (hunk 3 above)
Comment (TrevorDecker) [inline real_time/surface_src/java_src/Alice/src/main/java/com/roboclub/robobuggy/main/mainFile.java]: Codacy Issue found: [Avoid unused local variables such as 'LoggingKey'.](https://www.codacy.com/app/Mr-Trevor-Decker/RoboBuggy/file/1273313785/issues/source?bid=2938665&fileBranchId=2938743#l63)

File: real_time/surface_src/java_src/Alice/src/main/java/com/roboclub/robobuggy/nodes/GpsNode.java [hunk 11]
```diff
@@ -3,8 +3,12 @@
...
```
Comment (TrevorDecker) [inline real_time/surface_src/java_src/Alice/src/main/java/com/roboclub/robobuggy/nodes/GpsNode.java]: Codacy Issue found: [Avoid unused imports such as 'jdk.nashorn.internal.scripts.JS'](https://www.codacy.com/app/Mr-Trevor-Decker/RoboBuggy/file/1273313807/issues/source?bid=2938665&fileBranchId=2938743#l6)

File: real_time/surface_src/java_src/Alice/src/main/java/com/roboclub/robobuggy/main/config.java [hunk 12]
```diff
@@ -26,6 +26,23 @@
...
+	public static final String SENSOR_NAME_CAMERA = "camera";
+	public static final String SENSOR_NAME_ENCODER = "encoder";
+	public static final String SENSOR_NAME_GPS = "gps";
+	public static final String SENSOR_NAME_GUI_LOGGING_BUTTON = "logging_button";
+	public static final String SENSOR_NAME_IMU = "imu";
+	public static final String SENSOR_NAME_STEERING = "steering";
+	
+	//other message types
+	public static String MESSAGE_TYPE_RESET = "reset";
+	public static String MESSAGE_TYPE_STATUS = "status";
```
Comment (TrevorDecker) [inline real_time/surface_src/java_src/Alice/src/main/java/com/roboclub/robobuggy/main/config.java]: Codacy Issue found: [The field name indicates a constant but its modifiers do not](https://www.codacy.com/app/Mr-Trevor-Decker/RoboBuggy/file/1273313784/issues/source?bid=2938665&fileBranchId=2938743#l44)

**Comment by VivaanBahl** [comment]
@TrevorDecker commit 28cae92 was tested, and a sensors are logging properly. I'm currently optimizing the system to take advantage of messages, so any newer commits are broken

**Comment by TrevorDecker** [comment]
@VivaanBahl You should get this commit to pass travis, if you are going to make changes that you do not want in this pull request then you should make a new branch.

File: real_time/surface_src/java_src/Alice/src/main/java/com/roboclub/robobuggy/logging/SensorLogger.java [hunk 13]
```diff
@@ -9,40 +9,124 @@
...
+						stream.println("        " + parseData(message) + ",");
 					} catch (InterruptedException e) {
 						//TODO add to messages 
 						e.printStackTrace();
 					}
 				}
 			}
+
+			private String parseData(Message message) {
+				// TODO Auto-generated method stub
```
Comment (TrevorDecker) [inline real_time/surface_src/java_src/Alice/src/main/java/com/roboclub/robobuggy/logging/SensorLogger.java]: remove this comment

File: real_time/surface_src/java_src/Alice/src/main/java/com/roboclub/robobuggy/messages/BrakeMessage.java [hunk 14]
```diff
@@ -54,4 +55,10 @@ public Message fromLogString(String str) {
 		boolean brake_state = Boolean.parseBoolean(spl[1]);
 		return new BrakeMessage(d, brake_state);
 	}
+
+	@Override
+	public String getCorrespondingSensor() {
+		// TODO Auto-generated method stub
```
Comment (TrevorDecker) [inline real_time/surface_src/java_src/Alice/src/main/java/com/roboclub/robobuggy/messages/BrakeMessage.java]: Remove this comment

**Comment by VivaanBahl** [comment]
@TrevorDecker Sorry about that, didn't realize it would get into the pull request. I'm working on moving those commits elsewhere

**Comment by VivaanBahl** [comment]
@TrevorDecker Travis and buggy tests pass, can you do a final sweep before merging?

File: real_time/surface_src/java_src/Alice/src/main/java/com/roboclub/robobuggy/logging/SensorLogger.java (hunk 1 above)
Comment (TrevorDecker) [inline real_time/surface_src/java_src/Alice/src/main/java/com/roboclub/robobuggy/logging/SensorLogger.java]: Codacy Issue found: [Avoid unused imports such as 'com.roboclub.robobuggy.main.config'](https://www.codacy.com/app/Mr-Trevor-Decker/RoboBuggy/file/1278054617/issues/source?bid=2938665&fileBranchId=2938743#l13)

File: real_time/surface_src/java_src/Alice/src/main/java/com/roboclub/robobuggy/main/config.java (hunk 2 above)
Comment (TrevorDecker) [inline real_time/surface_src/java_src/Alice/src/main/java/com/roboclub/robobuggy/main/config.java]: Codacy Issue found: [The field name indicates a constant but its modifiers do not](https://www.codacy.com/app/Mr-Trevor-Decker/RoboBuggy/file/1278054624/issues/source?bid=2938665&fileBranchId=2938743#l29)

File: real_time/surface_src/java_src/Alice/src/main/java/com/roboclub/robobuggy/main/mainFile.java (hunk 3 above)
Comment (TrevorDecker) [inline real_time/surface_src/java_src/Alice/src/main/java/com/roboclub/robobuggy/main/mainFile.java]: Codacy Issue found: [Avoid unused local variables such as 'LoggingKey'.](https://www.codacy.com/app/Mr-Trevor-Decker/RoboBuggy/file/1278054625/issues/source?bid=2938665&fileBranchId=2938743#l63)

File: real_time/surface_src/java_src/Alice/src/main/java/com/roboclub/robobuggy/nodes/GpsNode.java (hunk 4 above)
Comment (TrevorDecker) [inline real_time/surface_src/java_src/Alice/src/main/java/com/roboclub/robobuggy/nodes/GpsNode.java]: Codacy Issue found: [Avoid unused imports such as 'jdk.nashorn.internal.scripts.JS'](https://www.codacy.com/app/Mr-Trevor-Decker/RoboBuggy/file/1278054648/issues/source?bid=2938665&fileBranchId=2938743#l6)

File: real_time/surface_src/java_src/Alice/src/main/java/com/roboclub/robobuggy/nodes/RBSMNode.java (hunk 5 above)
Comment (TrevorDecker) [inline real_time/surface_src/java_src/Alice/src/main/java/com/roboclub/robobuggy/nodes/RBSMNode.java]: Codacy Issue found: [Position literals first in String comparisons](https://www.codacy.com/app/Mr-Trevor-Decker/RoboBuggy/file/1278054652/issues/source?bid=2938665&fileBranchId=2938743#l163)

**Comment by TrevorDecker** [comment]
Looks good, can you pull this in before rolls tomorrow? 

If you have time aka its more important that we get this in to master for rolls. 
 I think that some of the auto style comments are actually kind of useful and should be applied
I also don't like keeping a lot of folders in the Alice directory can you move the schema files to a different folder.  I think the log files folder would make sense otherwise we may want to create a resource folder.

**Pull request merged**